### ISSUE TITLE
Release Self-Distilled Reasoner checkpoints on Hugging Face

### ISSUE THREAD

**Issue by NielsRogge** [siyan-zhao/OPSD#6]
Hi @siyan-zhao 🤗

I'm Niels and work as part of the open-source team at Hugging Face. I discovered your work through Hugging Face's daily papers as yours got featured: https://huggingface.co/papers/2601.18734.
The paper page lets people discuss about your paper and lets them find artifacts about it (your models for instance), you can also claim the paper as yours which will show up on your public profile at HF, and add Github/project page URLs.

Would you like to host the model checkpoints you've pre-trained using OPSD (such as the Self-Distilled Qwen3 variants) on https://huggingface.co/models? I see the code is already available on GitHub, but the pre-trained weights aren't linked yet.

Given the impressive reasoning performance you've achieved (especially on AIME), hosting these weights on Hugging Face would significantly improve their visibility and discoverability for the research community. We can add metadata tags so people can find the models easier, and link them directly to your paper page.

If you're down, leaving a guide [here](https://huggingface.co/docs/hub/models-uploading). If they are based on the Transformers/TRL stack, you can easily use the `push_to_hub` method. Alternatively, if it's a custom PyTorch model, you can use the [PyTorchModelHubMixin](https://huggingface.co/docs/huggingface_hub/package_reference/mixins#huggingface_hub.PyTorchModelHubMixin) class which adds `from_pretrained` and `push_to_hub` to the model.

After uploaded, we can also link the models to the paper page (read [here](https://huggingface.co/docs/hub/en/model-cards#linking-a-paper)) so people can discover your work.

You can also build a demo for your reasoning model on [Spaces](https://huggingface.co/spaces); we can provide you a ZeroGPU [grant](https://huggingface.co/docs/hub/en/spaces-gpus#community-gpu-grants), which gives you A100 GPUs for free.

Let me know if you're interested/need any guidance :)

Kind regards,

Niels
ML Engineer @ HF 🤗

**Comment by TongLiu-github** [comment]
+1

**Comment by penguinwang96825** [comment]
Hi @NielsRogge and @siyan-zhao 

I am not one of the authors of the OPSD paper/repository, but I have also been trying to reproduce the Qwen3-1.7B OPSD results reported in the paper. I trained the model using the authors’ code with the forward KL objective, using what I believe are the same hyperparameters as described in the paper, and evaluated the resulting LoRA checkpoints on AIME24, AIME25, and HMMT25.

However, my reproduced results do not exactly match the figures reported in the paper. The differences seem larger than a very small random fluctuation, so I wanted to share the details here in case it is useful for the discussion, and also to ask whether there are any additional settings or checkpoints needed to reproduce the reported numbers.

## Evaluation setup

For evaluation, I used the following setting:

```bash
BASE_MODEL="Qwen/Qwen3-1.7B"
EXP_DIR="/scratch/opsd_output/qwen3_1b_fkl_100steps"

NCCL_P2P_DISABLE=1 CUDA_VISIBLE_DEVICES=0,1,2,3 python eval/evaluate_math.py \
    --base_model "$BASE_MODEL" \
    --dataset "aime24" \
    --val_n 12 \
    --temperature 1.0 \
    --tensor_parallel_size 4

wait 

for checkpoint_path in "$EXP_DIR"/checkpoint-[0-9]*; do
    checkpoint_name=$(basename "$checkpoint_path")
    step=${checkpoint_name#checkpoint-}
    
    NCCL_P2P_DISABLE=1 CUDA_VISIBLE_DEVICES=0,1,2,3 python eval/evaluate_math.py \
        --base_model "$BASE_MODEL" \
        --dataset "aime24" \
        --val_n 12 \
        --temperature 1.0 \
        --tensor_parallel_size 4 \
        --checkpoint_dir "$checkpoint_path"
done
```

I used the same evaluation configuration for `aime25` and `hmmt25`, changing only the `--dataset` argument.

## Training setup

For training, I used the following command:

```bash
MODEL_PATH="Qwen/Qwen3-1.7B"
OUTPUT_DIR="/scratch/opsd_output"

accelerate launch \
    --config_file accelerate.yaml \
    --num_processes 4 \
    --gradient_accumulation_steps 4 \
    --main_process_port 12949 \
    opsd_train.py \
    --model_name_or_path "$MODEL_PATH" \
    --learning_rate 5e-6 \
    --max_grad_norm 0.1 \
    --per_device_train_batch_size 2 \
    --gradient_checkpointing \
    --gradient_accumulation_steps 4 \
    --output_dir "$OUTPUT_DIR" \
    --run_config qwen3_1b_fkl_100steps \
    --max_steps 100 \
    --max_completion_length 1024 \
    --save_steps 25 \
    --logging_steps 1 \
    --attn_implementation flash_attention_2 \
    --torch_dtype bfloat16 \
    --max_length 20000 \
    --beta 0 \
    --use_vllm \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.6 \
    --vllm_tensor_parallel_size 1 \
    --use_peft \
    --lora_r 64 \
    --lora_alpha 128 \
    --lora_target_modules q_proj k_proj v_proj o_proj gate_proj up_proj down_proj \
    --temperature 1.1 \
    --top_p 0.95 \
    --top_k 20 \
    --lmbda 1 \
    --fixed_teacher \
    --jsd_token_clip 0.05 \
    --wandb_project opsd
```

## Results comparison

Below is a combined summary comparing the paper-reported Qwen3-1.7B OPSD results with my reproduced evaluation results.

The table reports `Avg@12`, since this seems to be the closest metric to the paper table.

| Source / Model | Checkpoint | AIME24 Avg@12 | AIME25 Avg@12 | HMMT25 Avg@12 | Mean |
| :--- | :---: | ---: | ---: | ---: | ---: |
| Paper | OPSD Qwen3-1.7B | 57.20 | 43.90 | 29.20 | 43.40 |
| Reproduced | Base Qwen3-1.7B | 46.94 | 36.67 | 23.33 | 35.65 |
| Reproduced | OPSD ckpt-25 | **54.17** | 41.11 | 25.56 | 40.28 |
| Reproduced | OPSD ckpt-50 | 50.56 | 41.67 | **26.39** | 39.54 |
| Reproduced | OPSD ckpt-75 | 53.89 | **45.28** | 23.61 | 40.93 |
| Reproduced | OPSD ckpt-100 | 53.06 | 42.78 | 25.28 | 40.37 |
| Reproduced | Best per dataset | **54.17** | **45.28** | **26.39** | **41.95** |

For completeness, here are the full metrics from my reproduced evaluation runs:

| Dataset | Source / Model | Checkpoint | Pass@12 | Avg@12 | Majority Vote@12 | Format Rate |
| :--- | :--- | :---: | ---: | ---: | ---: | ---: |
| AIME24 | Reproduced | Base | 76.67 | 46.94 | 66.67 | 96.94 |
| AIME24 | Reproduced | OPSD ckpt-25 | **80.00** | **54.17** | 66.67 | **99.44** |
| AIME24 | Reproduced | OPSD ckpt-50 | 73.33 | 50.56 | 66.67 | **99.44** |
| AIME24 | Reproduced | OPSD ckpt-75 | 76.67 | 53.89 | 63.33 | 98.61 |
| AIME24 | Reproduced | OPSD ckpt-100 | 76.67 | 53.06 | **73.33** | 98.06 |
| AIME25 | Reproduced | Base | 66.67 | 36.67 | 50.00 | 98.61 |
| AIME25 | Reproduced | OPSD ckpt-25 | 66.67 | 41.11 | **56.67** | 99.17 |
| AIME25 | Reproduced | OPSD ckpt-50 | 66.67 | 41.67 | 53.33 | **99.72** |
| AIME25 | Reproduced | OPSD ckpt-75 | **66.67** | **45.28** | **56.67** | 99.17 |
| AIME25 | Reproduced | OPSD ckpt-100 | 63.33 | 42.78 | 53.33 | **99.72** |
| HMMT25 | Reproduced | Base | **60.00** | 23.33 | 26.67 | 98.89 |
| HMMT25 | Reproduced | OPSD ckpt-25 | 56.67 | 25.56 | 30.00 | **100.00** |
| HMMT25 | Reproduced | OPSD ckpt-50 | 56.67 | **26.39** | 30.00 | **100.00** |
| HMMT25 | Reproduced | OPSD ckpt-75 | 36.67 | 23.61 | 26.67 | **100.00** |
| HMMT25 | Reproduced | OPSD ckpt-100 | 50.00 | 25.28 | **33.33** | 99.44 |

From these runs, the best reproduced checkpoints by `Avg@12` are:

- **AIME24:** ckpt-25, 54.17 vs. 57.20 reported in the paper
- **AIME25:** ckpt-75, 45.28 vs. 43.90 reported in the paper
- **HMMT25:** ckpt-50, 26.39 vs. 29.20 reported in the paper

If anyone is interested in these reproduced LoRA checkpoints, I would be happy to upload them to the HuggingFace Hub as well.

Warm regards,
Yang

**Comment by siyan-zhao** [comment]
Hi @penguinwang96825, thanks a lot for sharing your reproduced results. I also noticed that the variance across different runs can be high. My reported results were obtained by running the released code on H100 GPUs, but I realize the current reporting may not make this variance clear enough. For a future version I will report results over multiple seeds to make the results more reliable. For reference, https://github.com/MBZUAI-reasoninglab/OP2SD reports OPSD reproduced results across 4 seeds. Hope this references can help, and sorry for the inconvenience and confusion!
